Add yesNo react component

JulianRoesner opened the pull request:
# Short Description
To reactify the StepSteuernummer we need to add some components in React. One is the YesNo component. I based the design on this [document](https://www.figma.com/file/sEwfGWO9XWlJ51bQ2ajBkl/Steuerlotse-Components?node-id=3639%3A5085) in Figma and talked to Nadine. It adapts also a little how the YesNo component behaves because we had some small legacy 

# Changes
- Add YesNo component
- Add YesNo story
- Adapt the switch-yes/switch-no classes a little bit

# Feedback
- Does this make sense to you?
- Is there anything that I missed while reactifying?

nfelger commented on webapp/client/src/components/FormFieldYesNo.js:
```diff
@@ -0,0 +1,162 @@
+  required: false,
+  autofocus: false,
+  details: FieldLabelForSeparatedFields.defaultProps.details,
+  onChangeHandler: () => {},
```
There's a code smell flagged for this… First I didn't agree, but thinking it over I now agree it'd be better to default this to `undefined`. The code should work without any other modification and it's clearer this way that nothing has been set for this.

JulianRoesner replied:
I agree. At some point I had a problem with using `undefined` as a defaultProps for a function. That's why I did it like this because this was recommended in However, as I can not reproduce this I have changed it to `undefined` 👍

nfelger commented on webapp/client/src/components/FormFieldDropDown.js:
```diff
@@ -0,0 +1,114 @@
+import PropTypes from "prop-types";
```
I wouldn't have expected the DropDown to be added here? I'll assume that's an accidental commit and ignore it?

JulianRoesner replied:
Yes. I'll remove it with the next commit

nfelger commented on webapp/client/src/components/FormFieldYesNo.js:
```diff
@@ -0,0 +1,162 @@
+      hideLabel
+      render={() => (
+        <YesNo>
+          <FieldLabelForSeparatedFields {...{ label, fieldId, details }} />
```
Interesting. Just going to note here that with `FieldLabelForSeparatedFields` and `FieldLabelScaffolding`, we've overloaded the term "label" again and would be better off using a different word, _especially_ for separated fields (and radio buttons) where this "FieldLabel" isn't rendering any label at all. But that's out of scope for this PR, I just wanted to highlight it.

nfelger replied:
No, it's fine as it is. I intended `FieldLabelScaffolding` (not a great name - open to suggestions!) to hold what's shared between both types of "Labels" that we have, not to be used on its own.

nfelger commented on webapp/client/src/components/FormFieldYesNo.js:
```diff
@@ -0,0 +1,162 @@
+          <fieldset id={fieldId} name={fieldId}>
+            <label
+              htmlFor={yesFieldId}
+              className={`switch-yes ${
```
https://www.npmjs.com/package/classnames makes conditionally setting classes much easier. See here for an example: https://github.com/digitalservice4germany/steuerlotse/blob/650042b632a6fffb9979f2e1b53d36456e779604/webapp/client/src/components/FormFieldScaffolding.js#L49

JulianRoesner replied:
Thanks that is a great idea and is so much better.

nfelger commented on webapp/client/src/stories/FormFieldDropDown.stories.js:
```diff
@@ -0,0 +1,49 @@
+import React from "react";
```
Also an accidental commit?

JulianRoesner replied:
Yep (s. above)

nfelger commented on webapp/client/src/stories/FormFieldYesNo.stories.js:
```diff
@@ -0,0 +1,86 @@
+  },
+  errors: [],
+  value: "no",
+  onChangeHandler: (event) => {
```
I like having an example with a change handler!

nfelger reviewed (changes requested):
Looks great, but I think this component lacks unit tests for:
- checking that toggling works as intended
- checking that the change handler is called as expected

See here for an example test suite https://github.com/digitalservice4germany/steuerlotse/blob/sqlachemy-pool-tweaks/webapp/client/src/components/FormFieldSeparatedField.test.js#L0-L1

NB: I didn't review the CSS and didn't look at it locally. Let me know if you'd like me to!

JulianRoesner commented:
> Looks great, but I think this component lacks unit tests for:
> 
> - checking that toggling works as intended
> -  checking that the change handler is called as expected


I agree. I have added these tests and I have also added tests for the interaction via the keyboard as recommended in #336 because I think it makes sense here, too. I had to update the testing library for that.

nfelger commented on webapp/client/src/components/FormFieldYesNo.js:
```diff
@@ -95,38 +104,30 @@ function FormFieldYesNo({
       render={() => (
         <YesNo>
-          <FieldLabelForSeparatedFields {...{ label, fieldId, details }} />
+          <FieldLabel {...{ label, fieldId, details }} />
```
Wait, I'm confused. I thought this _should_ use `FieldLabelForSeparatedFields`? [In the old world](https://github.com/digitalservice4germany/steuerlotse/blob/650042b632a6fffb9979f2e1b53d36456e779604/webapp/app/templates/components.html#L110), RadioFields (of which YesNo is a subclass) get the "Label" which isn't a label but a legend and that's what `FieldLabelForSeparatedFields` represents.

JulianRoesner replied:
I have changed it because we currently display a label. However, you are correct. It should be a legend and not a label because it "labels" a fieldset.

I think that this should probably be the direction we take for naming the different FieldLabel components (label, legend...)

nfelger commented on webapp/client/src/components/FormFieldYesNo.test.js:
```diff
@@ -0,0 +1,119 @@
+
+describe("FormFieldYesNo", () => {
+  let props;
+  const onChangeHandler = jest.fn();
```
Didn't know about these! Love it.

nfelger commented on webapp/client/src/components/FormFieldYesNo.test.js:
```diff
@@ -0,0 +1,119 @@
+
+  describe("When Ja clicked", () => {
+    beforeEach(() => {
+      fireEvent.click(screen.getByText("Ja"));
```
What's your reasoning for using `fireEvent` over `userEvent` here? I don't fully understand the difference, but the docs for `fireEvent` say:

> Most projects have a few use cases for fireEvent, but the majority of the time you should probably use @testing-library/user-event.

JulianRoesner replied:
Good point. I started with fireEvent before using the userEvent library and just kept it. I have changed this.

nfelger reviewed (approved):
❤️❤️❤️ for the keyboard testing! Have a look whether you've really used the correct "Label" component here, otherwise this is great!

nfelger commented:
Oh, and I agree on your assessment re: duplication. Any idea I would have for deduplicating those tests would make them less readable and maintainable, not more.

nfelger commented:
I reconfigured SonarCube to exclude JS tests from duplication checks, so your next commit should resolve the quality gate failure - let me know if it doesn't!
